subsys/net/ip/net_if.c

```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
 		 */
 		net_ipv6_nbr_rm(iface, &ifaddr->address.in6_addr);
 	}
+
+	ifaddr->dad.timer_timeout = 0;
+
+	sys_slist_find_and_remove(&active_dad_timers, &ifaddr->dad.node);
+}
+
+static bool dad_check_timeout(s64_t start, u32_t time, s64_t timeout)
+{
+	start += time;
+	start = abs(start);
+
+	if (start > timeout) {
+		return false;
+	}
+
+	return true;
+}
+
+static bool dad_timedout(struct net_if_addr *ifaddr, s64_t timeout)
+{
+	return dad_check_timeout(ifaddr->dad.timer_start,
+				 ifaddr->dad.timer_timeout,
+				 timeout);
+}
+
+static u32_t dad_manage_timeouts(struct net_if_addr *ifaddr, s64_t timeout)
+{
+	s32_t next_timeout;
+
+	if (dad_timedout(ifaddr, timeout)) {
+		dad_expired(ifaddr);
+		return UINT32_MAX;
+	}
+
+	next_timeout = timeout - (ifaddr->dad.timer_start +
+				  ifaddr->dad.timer_timeout);
+	return abs(next_timeout);
+}
+
+static void dad_timeout(struct k_work *work)
+{
+	u32_t timeout_update = UINT32_MAX - 1;
+	s64_t timeout = k_uptime_get();
+	struct net_if_addr *current, *next;
+
+	ARG_UNUSED(work);
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&active_dad_timers,
+					  current, next, dad.node) {
+		u32_t next_timeout;
+
+		next_timeout = dad_manage_timeouts(current, timeout);
+		if (next_timeout < timeout_update) {
+			timeout_update = next_timeout;
+		}
+	}
+
+	if (timeout_update < (UINT32_MAX - 1)) {
+		NET_DBG("Waiting for %u ms", timeout_update);
+
+		k_delayed_work_submit(&address_lifetime_timer, timeout_update);
```

Comment:
It's a bit strange that a routine related to DAD, with dad_timer available, submits work against address_lifetime_timer. Please add a code comment explaining why.

Reply:
#2 issue: you pass unsigned value to the function which expects signed one (k_delayed_work_submit). The right solution is of course to fix k_delayed_work_submit(), which takes s32_t, but deals with it as if it was u32_t (violating C standard in regard to overflow handling, etc.).